### ISSUE TITLE
Zipkin trace doesn't open/close span correctly?

### ISSUE THREAD

**Issue by kojilin** [line/armeria#811]
I found that sometimes my code has unpredictable `currentSpan` in some thread. I suspect that there is nested makeContextAware and cause unbalanced zipkin span scope open/close. Should we implement SpanContextUtil like below?

```
    public static void setupContext(RequestContext ctx, Span span, Tracer tracer) {
        ctx.onEnter(unused -> {
            Deque<SpanInScope> queue = ctx.attr(SPAN_IN_SCOPE_KEY).get();
            if (queue == null) {
                queue = new ArrayDeque<>();
                ctx.attr(SPAN_IN_SCOPE_KEY).set(queue);
            }
            queue.push(tracer.withSpanInScope(span));
        });
        ctx.onExit(unused -> {
            Deque<SpanInScope> spanInScopes = ctx.attr(SPAN_IN_SCOPE_KEY).get();
            if (spanInScopes != null) {
                SpanInScope currentSpan = spanInScopes.pop();
                if (currentSpan != null) {
                    currentSpan.close();
                }
            }
        });
    }
```
If onEnter & onExit will be called in pairs, we don't need to do null check.(and it should, right?)

**Comment by kojilin** [comment]
cc @trustin & @anuraaga

**Comment by anuraaga** [comment]
Hmm - I'm curious how you can end up with a double `makeContextAware` it seems like where a thread-switch was expected, there wasn't one.

Either way it seems reasonable to make `RequestContext` reentrant. I don't think we need the overhead of the queue, can just use a counter to track the enters and exits and only close on the last exit. It probably makes sense to handle this within `RequestContext` itself by adding `RequestContext.onFirstEnter / onLastExit`.

Another option would be to actually not allow `RequestContext` to be reentrant and make it an `IllegalStateException` to try to do so. We could add a reference to the parent to any child contexts, and traverse up during the context push to see if it's already been pushed. I can't think of a situation where nested `makeContextAware` would actually be necessary and think it's always a programming bug, where the user doesn't understand what thread some code is running on, so making it easy to find these by throwing seems ok to me too, but either option would work.

WDYT @trustin @kojilin ?

**Comment by kojilin** [comment]
you are right, let me check where I make it have onEnter twice.

**Comment by kojilin** [comment]
asked as slack, but add it here too.
If we submit runnable to non armeria executor with makeContextAware, `ctx.attr` will be overwritten?

**Comment by anuraaga** [comment]
As discussed offline, I was confused thinking reentrant push on the same thread, when it's actually very easy to trigger the behavior when the same ctx owns multiple context-aware callbacks run concurrently (e.g., on a thread-pool).

The queue will have an ordering issue, so I think it should be something like

```java
private static FastThreadLocal<SpanInScope>

onEnter(() -> threadLocal.set(tracer.withSpanInScope));
onExit(() -> threadLocal.get().close());
```

**Comment by kojilin** [comment]
thanks, my first suspect is wrong and make wrong solve at first. Create a PR and will add test if possible.

**Comment by trustin** [comment]
Fixed by #812
